binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java

```diff
@@ -193,6 +197,16 @@ ListenableFuture<Status> checkAuthorization(MethodDescriptor<?, ?> method) {
           serverPolicyChecker.checkAuthorizationForServiceAsync(uid, serviceName);
       if (useCache) {
         serviceAuthorization.putIfAbsent(serviceName, authorization);
+        Futures.addCallback(authorization, new FutureCallback<Status>() {
+          @Override
+          public void onSuccess(Status result) {}
+
+          @Override
+          public void onFailure(Throwable t) {
+            serviceAuthorization.remove(serviceName, authorization);
+            throw new IllegalStateException(t);
```

Comment:
Help me understand why you throw here? Your direct executor means this exception is going to happen on the thread that is trying to complete the future and report the failure. Worse, if the future has already failed, it's going to throw on the thread currently calling checkAuthorization().

Just calling remove() should be sufficient right?

Reply:
Ooops, you're absolutely right. This is a vestige from my previous attempt at using `Futures.catching` for retaining the failed aspect of the future.

Removed.

Reply:
Thanks. Can you update your unit test to fail in the presence of this bug? I think the problem is that you check for INTERNAL without evidence that the internal error comes from the failed security policy. Try failing the future with your own unique Exception subclass and asserting for that as the INTERNAL's cause.

Reply:
Done. I had to manually put the original exception's message into the `Status`' description though. According to the [writeStatus](https://github.com/grpc/grpc-java/blob/062f7a20722ba7519ad872b4c82bfc8259d8b7bc/binder/src/main/java/io/grpc/binder/internal/TransactionUtils.java#L83-L91) function, it doesn't seem like the status' caused is serialized, only its description and code. Otherwise, my test would get no info about the cause (i.e. the test just gets `Status{code=INTERNAL, description=null, cause=null}`.

Maybe a general library-wide improvement for the future is to serialize the cause as well?

Let me know if you have a better idea.